libsolidity/formal/CHC.cpp

```diff
@@ -323,6 +376,109 @@ void CHC::unknownFunctionCall(FunctionCall const&)
 	m_unknownFunctionCallSeen = true;
 }
 
+/*
+ * Loop encoding is described as follows.
+ * Let f_i the function body block before the loop.
+ * Create the following blocks:
+ * loop_header, containing constraints created by the condition (potential side effect statements).
+ * loop_body, containing constraints from the loop body.
+ * f_j, the function body block after the loop.
+ * Create the following edges:
+ * f_i -> loop_header
+ * loop_header -> loop_body
+ * loop_header -> f_j
+ * loop_body -> loop_header, if there were no nested loops inside this loop body.
+ * f_k -> loop_header, if there was at least one nested loop inside this loop body,
+ * where f_k is the function body block after the latest nested loop.
+ */
+void CHC::visitLoop(
+	BreakableStatement const& _loop,
+	Expression const* _condition,
+	Statement const& _body,
+	ASTNode const* _postLoop
+)
+{
+	solAssert(m_currentFunction, "");
+	auto const& functionBody = m_currentFunction->body();
+
+	// Create loop header block and edge from function body to loop header.
+	m_predicates[&_loop] = createBlock(sort(functionBody), "loop_" + to_string(_loop.id()));
+	smt::Expression loopHeader = predicateBodyCurrent(&_loop);
+	smt::Expression functionLoop = smt::Expression::implies(
+		m_path.back() && m_context.assertions(),
+		loopHeader
+	);
+	addRule(functionLoop, &functionBody, &_loop);
+
+	// Loop header block visits the condition and has out-edges
+	// to the loop body and back to the rest of the function.
+	pushBlock(loopHeader);
+
+	if (_condition)
+		_condition->accept(*this);
+	auto condition = _condition ? expr(*_condition) : smt::Expression(true);
+
+	// Create loop body entry block.
+	m_predicates[&_body] = createBlock(sort(functionBody), "loop_body_" + to_string(_body.id()));
+	// This predicate needs to be created here to take potential
+	// loop condition side-effects into account.
+	smt::Expression loopBody = predicateBodyCurrent(&_body);
+	smt::Expression bodyEdge = smt::Expression::implies(
+		loopHeader && m_context.assertions() && condition,
+		loopBody
+	);
+	addRule(bodyEdge, &_loop, &_body);
+
+	// Loop body block visit.
+	pushBlock(loopBody);
+
+	auto functionBlocks = m_functionBlocks;
+	_body.accept(*this);
+	// TODO handle properly when break/continue are supported.
+	if (_postLoop)
+		_postLoop->accept(*this);
+
+	// If there are nested inner loops new function blocks
+	// are created within this loop body.
+	// In that case the back edge needs to come from the function block.
```

Comment:
Is this really correct? It seems a bit off to me... Ah, the function body has SSA indices that are increased so that's why this will in fact be *after* the inner loops... ok, makes sense... maybe continue the comment with ``needs to come from the function block, which had its SSA index increased during the inner loop visit.`` or sth to make this clearer?